scalyr_agent/copying_manager/worker.py

```diff
@@ -964,15 +978,13 @@ def handle_completed_callback(result):
             # TODO:  This might not be bullet proof here.  We copy __log_processors and then update it at the end
             # We could be susceptible to exceptions thrown in the middle of this method, though now should.
 
-            # Copy the processor list because we may need to remove some processors if they are done.
-            processor_list = self.__log_processors[:]
-            self.__log_processors = []
-
             add_events_request.close()
 
             total_bytes_copied = 0
 
-            for processor in processor_list:
+            self._closed_files_checkpoints_lock.acquire()
+
+            for processor in self.__log_processors:
```

Comment:
For safety, I would surround this with a `try ... finally` where the lock release is done in the `finally` clause.

Reply:
Wow, that's a huge oversight, thanks.